### PR TITLE
created Airport and Plane classes, and gave Plane Class a 'land' method

### PR DESCRIPTION
**Pull request by barrygrubb**

### CODE REVIEW

File: spec/plane_spec.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+end
+
+context 'When the airport is full' do
+
+  it 'Flying plane will not be permitted to land at this airport.' do
+    weather = double(:stormy? => false)
+    default_capacity = 30
+    airport = double(:capacity => default_capacity, :landed_planes => [], :confirm_plane_landed => true, :landed_planes_history => [])
+    plane = Plane.new(airport)
+    30.times { airport.landed_planes << :plane }
```
Comment (jaxdid) [inline spec/plane_spec.rb]: Use constant instead of magic number.

File: spec/weather_spec.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  it 'Responds to the stormy? method.' do
+    expect(subject).to respond_to(:stormy?)
+  end
+
+  it 'Can return either True or False from the stormy? method.' do
+    result = []
+    500.times do
+      result << subject.stormy?
+    end
+    expect(result).to include(true) and expect(result).to include(false)
```
Comment (jaxdid) [inline spec/weather_spec.rb]: Unit tests should have a single expectation.

File: spec/weather_spec.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+describe Weather do
+
+  it 'Responds to the stormy? method.' do
+    expect(subject).to respond_to(:stormy?)
+  end
+
+  it 'Can return either True or False from the stormy? method.' do
+    result = []
+    500.times do
```
Comment (jaxdid) [inline spec/weather_spec.rb]: May want to use stubbing for true and false instead of 500.times (random)

File: spec/airport_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    expect(subject.instance_variable_get(:@landed_planes)).to be_a(Array)
+  end
+
+  it 'Adds the newly landed plane to the array of landed planes.' do
+    plane = double(:plane)
+    subject.landed_planes << plane
+    expect(subject.instance_variable_get(:@landed_planes)).to include(plane)
+  end
+
+  it 'Can receive the "confirm_plane_landed" command with 1 argument.' do
```
Comment (jaxdid) [inline spec/airport_spec.rb]: Possibly redundant `respond_to` expectation

File: spec/airport_spec.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  it 'Confirms that a newly landed plane has landed.' do
+  plane = double(:plane)
+  subject.landed_planes_history << plane
+  subject.landed_planes << plane
+  expect(subject.confirm_plane_landed(plane)).to eq(true)
+  end
+
+  it 'Can receive the "confirm_plane_taken_off" command with 1 argument.' do
+    expect(subject).to respond_to(:confirm_plane_taken_off).with(1).argument
```
Comment (jaxdid) [inline spec/airport_spec.rb]: Possibly redundant `respond_to` expectation

File: lib/airport.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  DEFAULT_CAPACITY = 30
+
+  def  initialize(capacity = DEFAULT_CAPACITY)
+    @landed_planes = []
+    @landed_planes_history = []
+    @capacity = capacity
+    @weather = Weather.new
+  end
+
+  def confirm_plane_landed(plane)
```
Comment (jaxdid) [inline lib/airport.rb]: Could be made private

File: lib/airport.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    fail 'That plane has never been at this airport.' if !plane_landed_at_airport?(plane)
+    plane_at_airport?(plane)
+  end
+
+  def confirm_plane_taken_off(plane)
```
Comment (jaxdid) [inline lib/airport.rb]: Could be made private

File: lib/plane.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
+class Plane
+
+  def initialize(airport)
+    fail 'There is no room at the airport.' if airport_full?(airport)
+    @landed = true
+    add_to_airport_list_of_landed_planes(airport)
+  end
+
+  def land(airport, weather = airport.weather)
```
Comment (jaxdid) [inline lib/plane.rb]: Could be made private

File: lib/plane.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    fail 'This plane has already landed.' if @landed
+    fail 'The weather is too stormy to land.' if weather.stormy?
+    fail 'There is no room at the airport.' if airport_full?(airport)
+    @landed = true
+    add_to_airport_list_of_landed_planes(airport)
+  end
+
+  def take_off(airport, weather = airport.weather)
```
Comment (jaxdid) [inline lib/plane.rb]: Could be made private?

File: lib/airport.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    fail 'That plane has never been at this airport.' if !plane_landed_at_airport?(plane)
+    !plane_at_airport?(plane)
+  end
+
+  private
+
+  def plane_at_airport?(plane)
+    @landed_planes.include?(plane)
```
Comment (jaxdid) [inline lib/airport.rb]: Call method instead of accessing instance variable directly

**Comment by mechacoach** [comment]
You had your airport_challenge reviewed by **Alain Lemaire**

### The good points are:

* Overview of project in the README
* Correct installation instructions in README
* `context` or `describe` blocks are used (where appropriate) to group tests
* The solution avoids vacuous tests
* There are a minimal number of untested paths in the code
* The code follows Ruby naming conventions, Readability is supported by consistent naming
* Guard clauses are used where appropriate, Booleans are returned implicitly where appropriate, removeallcommented-outcode: There is no commented-out code in the submission
* Classes have single responsibilities, Each methods has a single responsibility
* Code avoids strings where it should use symbols
* Command/Query naming is not an issue in the code
* The solution does not have any redundant code

### You should consider the following improvements:

* Use named subject with `described_class`
* Use `before` blocks to set up objects rather than repeat code
* Avoid multiple `expect`s in `it` block
* Handle randomness in tests
* Get rid of redundant `respond_to` expectations
* Remove all commented-out code
* Do not expose internal implementation
* Avoid magic numbers (e.g. on capacity)
* Improper use of attr_*

### Additional comments:
As discussed: extensive covering of edge cases and comprehensive testing. Nice!

see https://github.com/makersacademy/airport_challenge/blob/master/docs/review.md for more details

**Comment by mechacoach** [comment]
You had your airport_challenge reviewed by **Alain Lemaire**

### The good points are:

* Overview of project in the README
* Correct installation instructions in README
* `context` or `describe` blocks are used (where appropriate) to group tests
* The solution avoids vacuous tests
* There are a minimal number of untested paths in the code
* The code follows Ruby naming conventions, Readability is supported by consistent naming
* Guard clauses are used where appropriate, Booleans are returned implicitly where appropriate, removeallcommented-outcode: There is no commented-out code in the submission
* Classes have single responsibilities, Each methods has a single responsibility
* Code avoids strings where it should use symbols
* Command/Query naming is not an issue in the code
* The solution does not have any redundant code

### You should consider the following improvements:

* Use named subject with `described_class`
* Use `before` blocks to set up objects rather than repeat code
* Avoid multiple `expect`s in `it` block
* Handle randomness in tests
* Get rid of redundant `respond_to` expectations
* Remove all commented-out code
* Do not expose internal implementation
* Avoid magic numbers (e.g. on capacity)
* Improper use of attr_*

### Additional comments:
As discussed: extensive covering of edge cases and comprehensive testing. Nice!

see https://github.com/makersacademy/airport_challenge/blob/master/docs/review.md for more details

**Pull request closed without merging**